instrumentation/net/http/otelhttp/internal/request/body_wrapper.go

```diff
@@ -0,0 +1,72 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package request // import "go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp/internal/request"
+
+import (
+	"io"
+	"sync"
+)
+
+var _ io.ReadCloser = &BodyWrapper{}
+
+// BodyWrapper wraps a http.Request.Body (an io.ReadCloser) to track the number
+// of bytes read and the last error.
+type BodyWrapper struct {
+	io.ReadCloser
+	OnRead func(n int64) // must not be nil
+
+	mu   sync.Mutex
+	read int64
+	err  error
+}
+
+// NewBodyWrapper creates a new BodyWrapper.
+func NewBodyWrapper(rc io.ReadCloser, onRead func(int64)) *BodyWrapper {
```

Comment:
To help communicate what should be passed here can we name this `body`.

```suggestion
func NewBodyWrapper(body io.ReadCloser, onRead func(int64)) *BodyWrapper {
```